### ISSUE TITLE
Quasar leaks dead stack frames

### ISSUE THREAD

**Issue by exFalso** [corda/corda#989]
This bug is a bit elaborate and it would take some time to produce a small test case. The gist:

* Quasar uses two growing arrays to represent the fiber stack, one for metadata and primitive values and one for Objects.
* The arrays grow but never shrink. When a function returns all references in the corresponding dead frame *should* be nulled out in the Objects array.
* However in certain scenarios the nulling out does not happen properly. The stack pointer is set fine though, so this doesn't corrupt the quasar runtime.
* This means that we may end up in a scenario where the fiber holds references to objects in dead frames, which means that we may leak memory, but more importantly those frames will be serialised by Kryo during checkpointing (this is how the bug was found).

The precise circumstances under which the buggy nulling out happens are still unclear, but it seems to happen around tricky control flow including try blocks with multiple catches/throws inside the catches.

We should dedicate some time to produce a small test case and submit the issue upstream to PU. In the meantime we can write a custom serialiser for quasar stacks that follows through the frame metadata and only serialises live frames.

**Comment by mikehearn** [comment]
Alternatively, could we adjust the fiber suspend path to reallocate the stacks to be the right size, or to do extra nulling out at that point?

**Comment by exFalso** [comment]
@mikehearn That's a better approach I think. I found another issue with the dead frames, a fresh callchain may create frames that overlap with the dead one. Each frame in quasar has a corresponding unused slot in the object array (at the same index the frame metadata is in the primitive array), which may thus be filled with a dead reference. This means we cannot simply do resizing, we have to follow through the frame metadata and clean up those slots as well.

One other thing that comes to mind: I'm actually not sure what happens to uninitialised variables. In the standard java runtime they are initialised to null, but quasar may rely on the assumption that fresh frames always contain nulls, so it may skip the explicit null assignment - this would mean that uninitialised objects may be "initialised" with ones from an overlapped dead frame!

**Comment by mikehearn** [comment]
Alright, sounds like it's time to file more bugs upstream with PU.

**Comment by exFalso** [comment]
Finally managed to figure out the cause and reproduced the bug, it's not related to exceptions, and it probably happens quite often actually.

The nulling out happens in `co.paralleluniverse.fibers.Stack.java:popMethod`, which loops over the frame's objects based on the `slots` number that's bitmasked into the frame metadata.
This number comes from a parameter to the corresponding `pushMethod` call, this parameter in turn is passed from the instrumented bytecode.
The number is calculated in the constructor of  `co.paralleluniverse.fibers.instrument.InstrumentMethod$FrameInfo`, in particular it loops over `org.objectweb.asm.tree.analysis.Frame.getLocals()` to figure out whether a variable is an object or a primitive, increments their respective counter, and at the end takes the maximum of them. So far so good.
However when i stepped through the instrumentation of the faulty frame half of the locals had a null type, which means they don't count into these numbers!
Then I looked at the method's bytecode, there is a nice list of local variables at the end of each function definition with the corresponding stack offset. The locals that had a null type had a corresponding offset **that was reused** by another variable. This basically happens when there are two subsequent variable scopes, both declaring variables.

So in summary: If a method has subsequent variable scopes that reuse stack slots, then `objectweb` cannot infer the type of the slot, which in turn means that quasar doesn't count them into its `slots` number, which in turn means the frame will not be nulled out completely on function return. Phew!

I think a hacky workaround in quasar would be to add the count of the null typed locals to the current maximum(primitive,object)

**Comment by mikehearn** [comment]
The library is called asm actually rather than objectweb. It's pretty widely used for bytecode manipulation:

http://asm.ow2.org/asm50/javadoc/user/index.html
http://download.forge.objectweb.org/asm/asm4-guide.pdf

ASM provides a framework for the symbolic execution of bytecode, in the case of a re-used variable slot, I think it is supposed to find the common base type (and I don't think you can change the basic type of a slot). So the question is why does it return null here when the ASM docs don't mention this as a possibility. 

I suspect the answer lies in TypeInterpreter.merge(). I can see various cases that are not handled. Can you try enabling Quasar debug logging and see if anything is printed for these methods? e.g.

                if (typeV.getSort() != typeW.getSort()) {
                    db.log(LogLevel.DEBUG, "Array and non-array type can't be merged: %s %s", v, w);
                    return BasicValue.UNINITIALIZED_VALUE;
                }

Merging primitives and non-prims is likewise unhandled, without any debug logging. But I'm not sure that's valid bytecode. I'd have to check the JVMS.

Given the new info, can you put together a repro case?

**Comment by mikehearn** [comment]
BTW I don't think the workaround you suggest is the right approach here. Presumably it means some stuff would be missed on the fiber stack instead. Fixing the type inference logic to ensure that slots are always handled properly would be the right way to go here, I think.

**Comment by exFalso** [comment]
Linking corresponding PU issue https://github.com/puniverse/quasar/issues/282

**Comment by exFalso** [comment]
@mikehearn Well if the inference is fixed then a slot will presumably contain *all* types that may go in, which means quasar should increment both the primitive count and the object count - but this is equivalent to adding the non-inferable count to the end

**Comment by gendal** [comment]
Wow... simultaneously impressed and scared by this...

**Comment by mikehearn** [comment]
Hm, you may be right. The key question is whether this data is used for anything other than resetting stack slots. If it isn't then simply adding in nulls would probably help.

**Comment by richardhundt** [comment]
There's been some discussion on the Kotlin coroutines implementation project about serialization. It seems that Kryo can do it. Not sure if it's feasible to switch from Quasar at this point, and not sure if you guys are aware that this is possible, but here is a gist demonstrating it anyway:

https://gist.github.com/Restioson/fb5b92e16eaff3d9267024282cf1ed72

**Comment by mikehearn** [comment]
Thanks Richard. The underlying bug in Quasar this issue talks about is actually now fixed upstream, we're just waiting for time to update and check that we see the fix working.

The problem with relying on Kotlin coroutines (other than it being unavailable when we started and experimental even now) is that we need to support other JVM languages like Java too. Only Quasar is language neutral in that respect ... well, neutral-ish. I suspect dynamic languages like JS or Ruby won't work with it. Haskell dialects like Eta should. Scala, iirc had some odd glitch, but it's not fundamental and I think fixed by Quasar's Java 9 support.

Kotlin Coroutines have also had bugs. At this point I am not convinced Quasar is excessively buggy given the task it's doing, which is rather complicated (it's basically a state machine transpiler). We also have a commercial support contract for Quasar, and the JVM architects have added code to Java 9 specifically to support Quasar, so I have some basic degree of confidence that they won't be totally breaking it.

**Comment by richardhundt** [comment]
Thanks, Mike for taking the time to respond in detail and roger on doing it at the byte code level.

>We also have a commercial support contract for Quasar, and the JVM architects have added code to Java 9 specifically to support Quasar, so I have some basic degree of confidence that they won't be totally breaking it.

That's great to hear, and a really important piece of information for me. Serialised continuations aren't exactly a main-stream paradigm, so given that the core flow logic is built on them, it's good to know that the JVM and Quasar guys are working together, and that you guys have support. It matters when considering building a product on the Corda platform.

**Comment by mikehearn** [comment]
You are right. Perhaps we should advertise this fact more widely.

**Comment by richardhundt** [comment]
This is getting a bit off topic, and it's probably a bit niche, but I was wondering about the implications of being able to resume continuations more than once. I remember a conversation I had a few years back with the author of the Perl Coro library where I was trying to persuade him to add serialisation, and he pointed out that the world isn't idempotent, so being able to resume coroutines an unbounded number of times - which you can do once serialised - doesn't make sense in general.

It occurred to me that solving the double-spend problem, gives you a free solution here as well. So anyway, I think it's really elegant combining serialised continuations with a distributed ledger in this way. Deeply impressed.

**Comment by mikehearn** [comment]
The question of when to resume serialised coroutines is something that's cropped up throughout the history of this technique. The Seaside framework used serialised coroutines to represent paths through a web app and app developers had to take special precautions due to the existence of the browser back button i.e. your flow could go backwards.

In Corda coroutines/flows are resumed for only two reasons: something they were waiting for happened, or the process restarted. In the latter case, it's possible for execution to go backwards a small amount as the operation that was happening at the time the process quit gets retried. So flows and the world are expected to be idempotent between checkpoints. If you do something that changes the world, like send a message to some other system, then it's time to take another checkpoint. Ideally the act of changing the world is atomic with the act of checkpointing and we can do that in some cases but sometimes there will always be a bit of skew. MQ engines can help out with message deduplication for this sort of reason.

Whilst it's true that resuming serialised coroutines multiple times isn't something you'd normally want to do,  Corda doesn't do that in normal operation. The only case I can think of where you'd want to do something like this is if you do the first part of a computation on one machine, serialise it, then resume it on lots of machines, where the execution path diverges based on some shard ID read after resumption.

**Comment by richardhundt** [comment]
Yeah, Seaside is what got me thinking about this sort of stuff (at the time I was trying to "abstract away the statelessness of HTTP" using Perl). I get that in Corda the suspended coroutine isn't distributed, and its state is tightly controlled.

I was thinking of the trust environment, though. Either an adversary getting hold of a serialised coroutine which transfers money out of my account, or simply a dishonest participant fiddling with the system. If its state is perfectly indistinguishable between invocations, I'd want some guarantees that this can only happen once. This all becomes moot though, because the state transitions, once committed, seem to prevent it. I think it's a nice property.

ASIDE: I've been blown away by the sophistication of banking trojans like Zeus. There's even a variant (of Gozi, I think) which can replay mouse and keystroke dynamics to fool biometrics based on these, so that's made me a bit paranoid. I know it's a different attack surface, but these guys put in a lot of effort, so guarantees like that matter to me. Even under normal operations, exactly once MQ semantics are tricky to guarantee.

**Comment by mikehearn** [comment]
Yeah. Corda assumes the node-local database is secure. It's the users job to protect access to it. It's the same as every blockchain system in that respect. But you're right that the notary will eliminate duplicate re-attempts to advance the ledger twice.

There's some planning for what to do about things like Zeus, the SWIFT hackers etc in the tech white paper, in the client side signing section - might be worth checking out!
